juju/juju
# Buffer mongo log record insertions

## Description of change

- Refactor and simplify the state.DbLogger code.
- Introduce a state/logdb package, which I intend to move the existing DbLogger and friends into, but first things first. This package initially holds a BufferedLogger, which buffers log records and flushes to an underlying logger (e.g. DbLogger) when either the buffer capacity or a time interval are exceeded.
- Update state.DbLogger.Log to insert records in bulk.
- Update the "release" function returned by StatePool.Get, and StatePool.Remove, to report whether or not they remove/close the State.
- Update the apiserver logsink/logtransfer code to share a single DbLogger for all uses of a shared state pool entry, to cut down on the number of sockets used for logging.
- Update the apiserver logsink/logtransfer code to use state/logdb.BufferedLogger to buffer log writes. The parameters are currently hard-coded: a buffer capacity of 1024, and flush interval of 2 seconds. We can make these configurable as needed.

Note to reviewers: there are a few refactoring commits, which you may wish to review first. If it makes your life easier, I'll split this up - I just wanted to get it in front of people ASAP.

## QA steps

1. juju bootstrap localhost
2. juju add-machine
3. juju debug-log -m controller

Watch the logs roll in; perform some activity and observe that, unless there is a flurry of logs, debug-log will take ~2 seconds to update. The number of mongo sockets and insertions should be reduced from before. I added temporary instrumentation to prove this to myself, but it should be possible to eyeball the improvements via mongo stats with a larger deployment.

## Documentation changes

None.

## Bug reference

None.
